mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -566,8 +582,8 @@ class Word2VecModel private[spark] (
     wordList.zip(cosVec)
       .toSeq
       .sortBy(-_._2)
-      .take(num + 1)
-      .tail
+      .filter(tup => wordOpt.map(w => !w.equals(tup._1)).getOrElse(true) && tup._2 != 1.0d)
+      .take(num)
```

Comment:
We'd actually need `.take(num + 1).filter(...).take(num)` to properly handle the cases where the filter isn't rejecting anything.  I'm assuming the filter is fairly cheap but you're right that it doesn't make sense to do it any more than is necessary.